File: app/src/main/java/com/aero51/moviedatabase/repository/retrofit/TheMovieDbApi.java
package com.aero51.moviedatabase.repository.retrofit;

import androidx.lifecycle.LiveData;

import com.aero51.moviedatabase.repository.model.tmdb.credits.Actor;
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorImagesResponse;
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse;
import com.aero51.moviedatabase.repository.model.tmdb.credits.MovieCredits;
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieGenresResponse;
import com.aero51.moviedatabase.repository.model.tmdb.movie.MovieSearchResult;
import com.aero51.moviedatabase.repository.model.tmdb.movie.MoviesByGenrePage;
import com.aero51.moviedatabase.repository.model.tmdb.movie.PopularMoviesPage;
import com.aero51.moviedatabase.repository.model.tmdb.movie.NowPlayingMoviesPage;

import com.aero51.moviedatabase.repository.model.tmdb.movie.TopRatedMoviesPage;
import com.aero51.moviedatabase.repository.model.tmdb.movie.UpcomingMoviesPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.AiringTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.PopularTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TrendingTvShowsPage;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowGenresResponse;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowSearchResult;
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.TvShowsByGenrePage;
import com.aero51.moviedatabase.utils.ApiResponse;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface TheMovieDbApi {



    //region is optional
    @GET("discover/movie")
    Call<TopRatedMoviesPage> getTopRatedMovies(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region, @Query("vote_count.gte") Integer vote_count, @Query("sort_by") String sort_by, @Query("without_genres") Integer  without_genres);

    //region is optional
    @GET("discover/movie")
    Call<PopularMoviesPage> getPopularMovies(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region,@Query("vote_count.gte") Integer vote_count,@Query("sort_by") String sort_by);



    //region is optional
    @GET("movie/now_playing")
    Call<NowPlayingMoviesPage> getNowPlayingMovies(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region);

    //region is optional
    @GET("movie/upcoming")
    Call<UpcomingMoviesPage> getUpcomingMovies(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region);

    //region is optional
    @GET("tv/popular")
    Call<PopularTvShowsPage> getPopularTvShows(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region);

    @GET("discover/tv")
    Call<PopularTvShowsPage> getPopularTvShows(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("watch_region") String region, @Query("sort_by") String sort_by, @Query("vote_count.gte") Integer vote_count);

    //region is optional
    @GET("tv/on_the_air")
    Call<AiringTvShowsPage> getAiringTvShows(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("region") String region);

    @GET("trending/tv/week")
    Call<TrendingTvShowsPage> getTrendingTvShows(@Query("api_key") String api_key
            , @Query("page") Integer page);

    @GET("movie/{movie_id}/credits")
    Call<MovieCredits> getMovieCredits(@Query("api_key") String api_key, @Path("movie_id") Integer movie_id);

    @GET("movie/{movie_id}/credits")
    LiveData<ApiResponse<MovieCredits>> getLiveMovieCredits(@Path("movie_id") Integer movie_id, @Query("api_key") String api_key);

    @GET("person/{person_id}")
    LiveData<ApiResponse<Actor>> getLivePerson(@Path("person_id") Integer person_id, @Query("api_key") String api_key);

    // GET /person/{person_id}/images
    @GET("person/{person_id}/images")
    LiveData<ApiResponse<ActorImagesResponse>> getLivePersonImages(@Path("person_id") Integer person_id, @Query("api_key") String api_key);


    @GET("search/person")
    LiveData<ApiResponse<ActorSearchResponse>> getLivePersonSearch(@Query("api_key") String api_key, @Query("query") String person);

    @GET("search/person")
    Call<ActorSearchResponse> getPersonsSearch(@Query("api_key") String api_key, @Query("query") String person, @Query("page") Integer page);

    @GET("search/movie")
    Call<MovieSearchResult> getMoviesSearch(@Query("api_key") String api_key, @Query("query") String movieSearch, @Query("page") Integer page);

    @GET("search/tv")
    Call<TvShowSearchResult> getTvShowsSearch(@Query("api_key") String api_key, @Query("query") String movieSearch, @Query("page") Integer page);

    @GET("genre/movie/list")
    LiveData<ApiResponse<MovieGenresResponse>> getLiveMovieGenres(@Query("api_key") String api_key);

    @GET("genre/tv/list")
    LiveData<ApiResponse<TvShowGenresResponse>> getLiveTvGenres(@Query("api_key") String api_key);

    @GET("discover/movie")
    Call<MoviesByGenrePage> getMoviesByGenre(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("with_genres") Integer with_genres, @Query("sort_by") String sort_by, @Query("vote_count.gte") Integer vote_count);


    @GET("discover/tv")
    Call<TvShowsByGenrePage> getTvShowsByGenre(@Query("api_key") String api_key
            , @Query("page") Integer page, @Query("with_genres") Integer with_genres, @Query("sort_by") String sort_by, @Query("vote_count.gte") Integer vote_count,@Query("without_genres") Integer without_genre);




}

File: app/src/main/java/com/aero51/moviedatabase/repository/retrofit/TheMovieDbApi.kt
package com.aero51.moviedatabase.repository.retrofit

import androidx.lifecycle.LiveData
import com.aero51.moviedatabase.repository.model.tmdb.credits.Actor
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorImagesResponse
import com.aero51.moviedatabase.repository.model.tmdb.credits.ActorSearchResponse
import com.aero51.moviedatabase.repository.model.tmdb.credits.MovieCredits
import com.aero51.moviedatabase.repository.model.tmdb.movie.*
import com.aero51.moviedatabase.repository.model.tmdb.tvshow.*
import com.aero51.moviedatabase.utils.ApiResponse
import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface TheMovieDbApi {
    //region is optional
    @GET("discover/movie")
    fun getTopRatedMovies(@Query("api_key") api_key: String
                          , @Query("page") page: Int, @Query("region") region: String, @Query("vote_count.gte") vote_count: Int, @Query("sort_by") sort_by: String, @Query("without_genres") without_genres: Int): Call<TopRatedMoviesPage>

    //region is optional
    @GET("discover/movie")
    fun getPopularMovies(@Query("api_key") api_key: String
                         , @Query("page") page: Int, @Query("region") region: String, @Query("vote_count.gte") vote_count: Int, @Query("sort_by") sort_by: String): Call<PopularMoviesPage>

    //region is optional
    @GET("movie/now_playing")
    fun getNowPlayingMovies(@Query("api_key") api_key: String
                            , @Query("page") page: Int, @Query("region") region: String): Call<NowPlayingMoviesPage>

    //region is optional
    @GET("movie/upcoming")
    fun getUpcomingMovies(@Query("api_key") api_key: String
                          , @Query("page") page: Int, @Query("region") region: String): Call<UpcomingMoviesPage>

    //region is optional
    @GET("tv/popular")
    fun getPopularTvShows(@Query("api_key") api_key: String
                          , @Query("page") page: Int, @Query("region") region: String): Call<PopularTvShowsPage>

    @GET("discover/tv")
    fun getPopularTvShows(@Query("api_key") api_key: String
                          , @Query("page") page: Int, @Query("watch_region") region: String, @Query("sort_by") sort_by: String, @Query("vote_count.gte") vote_count: Int): Call<PopularTvShowsPage>

    //region is optional
    @GET("tv/on_the_air")
    fun getAiringTvShows(@Query("api_key") api_key: String
                         , @Query("page") page: Int, @Query("region") region: String): Call<AiringTvShowsPage>

    @GET("trending/tv/week")
    fun getTrendingTvShows(@Query("api_key") api_key: String
                           , @Query("page") page: Int): Call<TrendingTvShowsPage>

    @GET("movie/{movie_id}/credits")
    fun getMovieCredits(@Query("api_key") api_key: String, @Path("movie_id") movie_id: Int): Call<MovieCredits>

    @GET("movie/{movie_id}/credits")
    fun getLiveMovieCredits(@Path("movie_id") movie_id: Int, @Query("api_key") api_key: String): LiveData<ApiResponse<MovieCredits>>

    @GET("person/{person_id}")
    fun getLivePerson(@Path("person_id") person_id: Int, @Query("api_key") api_key: String): LiveData<ApiResponse<Actor>>

    // GET /person/{person_id}/images
    @GET("person/{person_id}/images")
    fun getLivePersonImages(@Path("person_id") person_id: Int, @Query("api_key") api_key: String): LiveData<ApiResponse<ActorImagesResponse>>

    @GET("search/person")
    fun getLivePersonSearch(@Query("api_key") api_key: String, @Query("query") person: String): LiveData<ApiResponse<ActorSearchResponse>>

    @GET("search/person")
    fun getPersonsSearch(@Query("api_key") api_key: String, @Query("query") person: String, @Query("page") page: Int): Call<ActorSearchResponse>

    @GET("search/movie")
    fun getMoviesSearch(@Query("api_key") api_key: String, @Query("query") movieSearch: String, @Query("page") page: Int): Call<MovieSearchResult>

    @GET("search/tv")
    fun getTvShowsSearch(@Query("api_key") api_key: String, @Query("query") movieSearch: String, @Query("page") page: Int): Call<TvShowSearchResult>

    @GET("genre/movie/list")
    fun getLiveMovieGenres(@Query("api_key") api_key: String): LiveData<ApiResponse<MovieGenresResponse>>

    @GET("genre/tv/list")
    fun getLiveTvGenres(@Query("api_key") api_key: String): LiveData<ApiResponse<TvShowGenresResponse>>

    @GET("discover/movie")
    fun getMoviesByGenre(@Query("api_key") api_key: String
                         , @Query("page") page: Int, @Query("with_genres") with_genres: Int, @Query("sort_by") sort_by: String, @Query("vote_count.gte") vote_count: Int): Call<MoviesByGenrePage>

    @GET("discover/tv")
    fun getTvShowsByGenre(@Query("api_key") api_key: String
                          , @Query("page") page: Int, @Query("with_genres") with_genres: Int, @Query("sort_by") sort_by: String, @Query("vote_count.gte") vote_count: Int, @Query("without_genres") without_genre: Int): Call<TvShowsByGenrePage>

    @GET("movie/{movie_id}/videos")
    fun getLiveVideosForMovie(@Path("movie_id")movie_id: Int,@Query("api_key") api_key: String): LiveData<ApiResponse<MovieVideosResponse>>


}